### PR TITLE
test: minor version upgrade e2e testing

### PR DESCRIPTION
**Pull request by cmwaters**
## Overview

Closes: #313

Ref: #2633

This PR introduces a test for checking compatibility and the running of upgrades of minor versions.

It reads all minor versions tagged in git. Each node begins on a random version.

Each node then individually performs upgrades, going down and spinning back up on the new randomly chosen version (can be downgrades as well)

Currently we only check that their is no app version mismatch and the network is able to continually build blocks

## Checklist

- [ ] New and updated code has appropriate documentation
- [ ] New and updated code has new and/or updated testing
- [ ] Required CI checks are passing
- [ ] Visual proof for any user facing features like CLI or documentation updates
- [ ] Linked issues closed with keywords

### CODE REVIEW

File: Makefile
Comment (rootulp) [inline Makefile:120]: [quesiton] `git tag -l` appears to list all versions. Does this env variable contain all possible versions and not just a particular version?

```suggestion
	@export E2E_VERSIONS=$(git tag --list)
```

File: test/e2e/simple_test.go
Comment (rootulp) [inline test/e2e/simple_test.go:30]: Answering my own question from above, I think this contains all possible versions

File: test/e2e/upgrade_test.go
Comment (rootulp) [inline test/e2e/upgrade_test.go:22]: [question] I think this boolean will evaluate to true even if `E2E=false`. Potentially: 

```suggestion
	if os.Getenv("E2E") != "true" {
```

File: test/e2e/upgrade_test.go
Comment (rootulp) [inline test/e2e/upgrade_test.go:34]: [question] does this test also include patch versions?

^ disregard. Answered my own question by modifying a unit test below. This does include patch versions in the major version.

File: test/e2e/versions_test.go
Comment (rootulp) [inline test/e2e/versions_test.go:26]: I think I identified a bug in the `Order` function based on this unit test:

```go
func TestOrderOfReleaseCandidates(t *testing.T) {
	versionsStr := "v1.0.0\nv1.0.0-rc0\nv1.0.0-rc1\n"
	versions := e2e.ParseVersions(versionsStr)
	versions.Order()
	assert.Equal(t, versions[0], e2e.Version{1, 0, 0, true, 0})
	assert.Equal(t, versions[1], e2e.Version{1, 0, 0, true, 1})
	assert.Equal(t, versions[2], e2e.Version{1, 0, 0, false, 0})
}
```

There is a special case that I think needs to be handled. The sorted order should be:
1. v1.0.0-rc0
2. v1.0.0-rc1
3. v1.0.0

but currently v1.0.0 gets sorted before the release candidates

**Review by rootulp — COMMENTED** [review]
Great work! When I run this locally, I encounter:

```
$ make test-e2e
--> Running end to end tests
=== RUN   TestE2ESimple
    simple_test.go:34: Running simple e2e test version latest
    simple_test.go:37:
        	Error Trace:	/Users/rootulp/git/rootulp/celestia/celestia-app/test/e2e/simple_test.go:37
        	Error:      	Received unexpected error:
        	            	retrieving the Kubernetes config: stat /Users/rootulp/.kube/config: no such file or directory
        	Test:       	TestE2ESimple
--- FAIL: TestE2ESimple (0.00s)
=== RUN   TestMinorVersionCompatibility
    upgrade_test.go:27: skipping e2e test: E2E_VERSION not set
--- SKIP: TestMinorVersionCompatibility (0.00s)
```

What prerequisites do contributors need to perform prior to running make test-e2e locally? It looks like I need some kube config and I'm not sure why the the e2e test is getting skipped

**Comment by cmwaters** [comment]
> What prerequisites do contributors need to perform prior to running make test-e2e locally? It looks like I need some kube config and I'm not sure why the the e2e test is getting skipped

Yes you need a .kube/config setup. You can ask samuel for it. Also you need the E2E environment variable to run the tests (I purposefully didn't want go test ./... to run these tests). It's best to just run `make test-e2e`

File: test/e2e/upgrade_test.go
Comment (cmwaters) [reply test/e2e/upgrade_test.go:22]: Yeah, I could even combine `E2E` and `E2E_VERSIONS`

File: test/e2e/versions_test.go
Comment (cmwaters) [reply test/e2e/versions_test.go:26]: Ok nice, let me fix that

File: test/e2e/upgrade_test.go
Comment (rootulp) [reply test/e2e/upgrade_test.go:22]: SGTM

File: Makefile
Comment (rootulp) [reply Makefile:120]: [nit] Makes sense. This is super nit picky but then I think plural `E2E_VERSIONS` is more appropriate

**Comment by rootulp** [comment]
1. Can we add the `.kube/config` to this repo and document where maintainers should place it on their local machines? Doesn't strictly have to be in this PR.
2. hmm I tried `make test-e2e` locally. I can retry after the env variable merging

File: Makefile
Comment (cmwaters) [reply Makefile:120]: Yeah I can change that

**Comment by rootulp** [comment]
> no because this shouldn't be public.

[blocking] in this PR can we please indicate that `make test-e2e` isn't expected to work for external contributors.

File: test/e2e/upgrade_test.go
Comment (evan-forbes) [inline test/e2e/upgrade_test.go:82]: does this function block until the node is back up?

File: test/e2e/upgrade_test.go
Comment (evan-forbes) [inline test/e2e/upgrade_test.go:84]: what is this wait for?

File: test/e2e/upgrade_test.go
Comment (evan-forbes) [inline test/e2e/upgrade_test.go:82]: what the reasoning behind the `i < len(versions)*2` and `testnet.Node(i % numNodes)` ? depending on the number of versions that we're testing, will we upgrade nodes differently?

File: test/e2e/upgrade_test.go
Comment (evan-forbes) [inline test/e2e/upgrade_test.go:79]: is skipping the first node also the mechanism that is testing that the versions are compatible?

File: test/e2e/upgrade_test.go
Comment (evan-forbes) [inline test/e2e/upgrade_test.go:32]: what is the mechanism that is detecting a failure or halt in one of the nodes?

File: test/e2e/upgrade_test.go
Comment (evan-forbes) [inline test/e2e/upgrade_test.go:62]: linking #1535 here, as we eventually want to test more than blobs for the compatibility tests

**Review by evan-forbes — COMMENTED** [review]
had some questions, but its super dope we're getting to use knuu for compatibility tests

File: test/e2e/upgrade_test.go
Comment (cmwaters) [reply test/e2e/upgrade_test.go:84]: To give some time between nodes going down and upgrading

File: test/e2e/upgrade_test.go
Comment (cmwaters) [reply test/e2e/upgrade_test.go:82]: I actually don't know. It calls something in knuu (cc @smuu)

Comment (cmwaters) [reply test/e2e/upgrade_test.go:82]: Well we want the amount of upgrades to be proportional to the amount of versions there are so there's a higher probability we test all permutations

File: test/e2e/upgrade_test.go
Comment (cmwaters) [reply test/e2e/upgrade_test.go:79]: I don't quite understand. We need the first node to always be up because we're constantly submitting transactions. If txsim can't submit a transaction it exits.

File: test/e2e/upgrade_test.go
Comment (cmwaters) [reply test/e2e/upgrade_test.go:32]: Yeah I was thinking we should have some assertion that all nodes are within the same height

File: test/e2e/upgrade_test.go
Comment (cmwaters) [reply test/e2e/upgrade_test.go:62]: Yeah, we definitely should expand on the possible tx patterns to cover all tx types

File: Makefile
Comment (rootulp) [inline Makefile:119]: [nit] the newline in this comment causes the second line to appear like a command.

Screenshot 2023-11-06 at 1 04 36 PM


```suggestion
## test-e2e: Run end to end tests via knuu. This command requires a kube/config file to configure kubernetes.
```

File: test/e2e/upgrade_test.go
Comment (rootulp) [inline test/e2e/upgrade_test.go:29]:
```suggestion
		t.Skip("skipping e2e test: E2E_VERSIONS not set")
```

**Review by rootulp — COMMENTED** [review]
When I run this locally, I observe:

1. [not blocking] verbose logging from test e2e simple. Can we please redirect that logging so that the test results are easier to read?
2. [blocking] TestMinorVersionCompatibility gets skipped. TBH not sure why b/c the Makefile looks correct but my shell doesn't ever have these values populated.

```shell
$ make test-e2e
--> Running end to end tests
=== RUN   TestE2ESimple
    simple_test.go:34: Running simple e2e test version latest
< verbose logging from test e2e simple> 
--- PASS: TestE2ESimple (318.32s)
=== RUN   TestMinorVersionCompatibility
    upgrade_test.go:29: skipping e2e test: E2E_VERSION not set
--- SKIP: TestMinorVersionCompatibility (0.00s)
=== RUN   TestVersionParsing
--- PASS: TestVersionParsing (0.00s)
=== RUN   TestFilterMajorVersions
--- PASS: TestFilterMajorVersions (0.00s)
=== RUN   TestOrder
--- PASS: TestOrder (0.00s)
=== RUN   TestOrderOfReleaseCandidates
--- PASS: TestOrderOfReleaseCandidates (0.00s)
PASS
ok  	github.com/celestiaorg/celestia-app/test/e2e	318.931s
```

```shell
$ echo $E2E_VERSIONS

$ echo $E2E_VERSION

```

File: test/e2e/upgrade_test.go
Comment (evan-forbes) [reply test/e2e/upgrade_test.go:81]: I see now that the test has been updated for each upgraded node to wait a few heights. I was curious what the mechanism was to ensure that different versions were running concurrently

**Comment by cmwaters** [comment]
> [not blocking] verbose logging from test e2e simple. Can we please redirect that logging so that the test results are easier to read?

Yeah I can remove that. That was initially just to let me know it was working but should be fine to silence them now

> [blocking] TestMinorVersionCompatibility gets skipped. TBH not sure why b/c the Makefile looks correct but my shell doesn't ever have these values populated.

Let me look into it

**Review by evan-forbes — DISMISSED** [review]
lgtm :+1:

**Review by evan-forbes — APPROVED** [review]

**Comment by cmwaters** [comment]
@rootulp do you want to try it now

File: test/e2e/simple_test.go
Comment (rootulp) [inline test/e2e/simple_test.go]: is this residual from debugging?

```suggestion
```

File: test/e2e/versions_test.go
Comment (rootulp) [inline test/e2e/versions_test.go:43]:
```suggestion
	versionsStr := "v1.0.0 v1.0.0-rc0 v1.0.0-rc1"
```

**Review by rootulp — COMMENTED** [review]
testing upgrade test now and it appears to be running 👍 

Update: I quit after 10 mins b/c it didn't seem to be making progress

```
$ make test-e2e
--> Running end to end tests
=== RUN   TestE2ESimple
    simple_test.go:25:
--- SKIP: TestE2ESimple (0.00s)
=== RUN   TestMinorVersionCompatibility
versions v1.0.0	v1.1.0	v1.2.0	v1.3.0    upgrade_test.go:40: Running minor version compatibility test versions v1.0.0	v1.1.0	v1.2.0	v1.3.0
    upgrade_test.go:58: Starting node node 0 version v1.1.0
    upgrade_test.go:58: Starting node node 1 version v1.3.0
    upgrade_test.go:58: Starting node node 2 version v1.0.0
    upgrade_test.go:58: Starting node node 3 version v1.2.0








^Csignal: interrupt
FAIL	github.com/celestiaorg/celestia-app/test/e2e	585.399s
make: *** [test-e2e] Error 1
```

File: test/e2e/simple_test.go
Comment (cmwaters) [reply test/e2e/simple_test.go]: Hmm I thought I removed that

Comment (cmwaters) [reply test/e2e/simple_test.go]: Ah yeah it has been removed

File: test/e2e/upgrade_test.go
Comment (smuu) [reply test/e2e/upgrade_test.go]: Yes the current implementation blocks until the node is back up

**Review by smuu — APPROVED** [review]
This looks great!

**Review by evan-forbes — APPROVED** [review]

**Review by rootulp — APPROVED** [review]

**Pull request merged**
